public/app/core/logs_model.ts

```diff
@@ -288,6 +293,7 @@ export function logSeriesToLogsModel(logSeries: DataFrame[]): LogsModel {
         raw: message,
         labels: series.labels,
         timestamp: ts,
+        uid: `${ts}_${seriesLabelsString}_${rowLabelsString}`,
```

Comment:
Assuming ts is unique, why adding labels to the uid? Multiple log streams === ts may actually be not unique?

Reply:
ts itself is not unique. I assume this is because promtail on each node will generate unique timestamps but they do not need to be unique across all nodes, on the other hand I assume each node have unique labels so labels + ts should be unique. @cyriltovena correct me if I am wrong here.

Reply:
@joe-elliott @slim-bean when changing to float64 this might not be true anymore but if you add the index of the row that should do it.

Reply:
@cyriltovena not sure what do you mean by that. If we change the timestamp from ISO string to float64 I do not thing that should change the uid in any way should it? Or do you mean because of possible loss of precision when converting float64 to number (when creating the JSON)?

Reply:
Possible loss yes

Reply:
@cyriltovena I assume this is going to be problem only for JS right? I would assume that go serialiser will generate a JSON string with the correct nanosecond number and the problem will come up when parsing that. If that is the case I think there are libraries in JS that will parse json with big numbers without precision loss, so when we come to that I assume we should use that (and the cast to string here or convert to milliseconds for date formatting).

Reply:
We need to find a good way to maintain nanosecond precision when it exists.  This is big deal for hi-resolution sensor data.   Chronograph uses [nano-date](https://github.com/jcgertig/nano-date).  I'm not sure what the best near-term solution is.  Keep the string version?

Reply:
I updated my suggestion here https://github.com/grafana/loki/issues/273 this is for the new endpoint that you should eventually move to when the API will be stable. This would make the new endpoint use `"1568215619994013000"` string as nanoseconds timestamp. You can then use nano-date project to parse them. You should still be able to ensure row unicity with this string.

Should you give a try to nano-date to see if you can parse the example in the issue.